Lifetimes: Updates to incorporate NLL (#101)
rust-lang-nursery/nomicon: src/lifetimes.md

# Lifetimes

Rust enforces these rules through *lifetimes*. Lifetimes are effectively
just names for areas of code somewhere in the program (you could also look at
them as time intervals, except that the beginning and end of the time is denoted
by when a line of code executes). Each reference, and anything that contains a

> reference, is tagged with a lifetime specifying the time it's valid for.

Reviewer: I'm not sure I like this time metaphor. A more apt one might be "paths of execution"?

# Lifetimes

Rust enforces these rules through *lifetimes*. Lifetimes are effectively
just names for areas of code somewhere in the program (you could also look at
them as time intervals, except that the beginning and end of the time is denoted
by when a line of code executes). Each reference, and anything that contains a
reference, is tagged with a lifetime specifying the paths of executions it's

> valid for.

Reviewer: This paragraph is now a bit too much of a mess of metaphors. Suggested rewrite:

```markdown
Rust enforces these rules through *lifetimes*. Lifetimes are named
regions of code that a reference must be valid for. Those regions
may be fairly complex, as they correspond to paths of execution
in the program. There may even be holes in these paths of execution, 
as it's possible to invalidate a reference as long as it's reinitialized
before it's used again. Types which contain references (or pretend to)
may also be tagged with lifetimes so that Rust can prevent them from
being invalidated as well.
```

Author: Not the first time that happened to me, I guess my brain just works mostly on metaphors most of the time :-|. Anyway, adopting your version, thank you.

to the compiler. However it does mean that several programs that are totally
correct with respect to Rust's *true* semantics are rejected because lifetimes
are too dumb.

# The area covered by a lifetime

The lifetime (sometimes called a *borrow*) is *alive* from the place it is
created to its last use. The borrowed thing needs to outlive only borrows that
are alive. This looks simple, but there are few subtleties.

> This compiles, because after printing `x`, it is no longer needed, so it doesn't

Reviewer 2: This "this" is quite far away from the example it refers to, so I was confused for a bit here. Could you move it closer, or else make the reference more explicit?

Author: I moved the paragraph below that to the very end, that might be a better place.

needed, so it doesn't matter if it is dangling or aliased (even though the
variable `x` *technically* exists to the very end of the scope).

```rust,edition2018
let mut data = vec![1, 2, 3];
let x = &data[0];
println!("{}", x);
// This is OK, x is no longer needed
data.push(4);
```

However, if the value has a destructor, the destructor is run at the end of the
scope. And running the destructor is considered a use ‒ obviously the last one.
So, this will *not* compile.

```rust,edition2018,compile_fail
#[derive(Debug)]

> struct X<'a>(&'a i32);

Reviewer: clarity nit: Can you rename this to Borrow?

impl Drop for X<'_> {
    fn drop(&mut self) {}
}

let mut data = vec![1, 2, 3];
let x = X(&data[0]);
println!("{:?}", x);
data.push(4);

> // Here, the destructor is run and therefore this'll fail to compile.

Reviewer: ```suggestion
// Here, the destructor is run and prevents the borrow from expiring
```
